src/controller.py

```diff
@@ -19,99 +18,107 @@ def __init__(self, input_file=None, random_state=None):
-            logging.info(
-                f'Active player: {player.name}')
+        logging.info(
+            'Cards drawn. Now starting infect phase.')
 
-            game.start_turn(player)
+        self.game.infect_city_phase()
 
-            while player.action_count:
-                logging.info(
-                    f'Actions left: {player.action_count}')
+        logging.info(
+            'Infect phase gone. Starting new turn.')
 
-                command = self.input()
-                if not command:
-                    continue
+        self._switch_player()
 
-                logging.debug(
-                    'Player action: ' + command)
-                command = command.split()
-
-                chosen_executor = None
-                for executor in COMMANDS:
-                    potential_executor = executor(game, player, self)
-                    if potential_executor.check_valid_command(command):
-                        chosen_executor = potential_executor
-                        break
-                if not chosen_executor:
-                    logging.error(
-                        'Command cannot be parsed. Type a correct command.')
-                    continue
+    def run_single_command(self, command):
+        logging.debug(
+            'Player action: ' + command)
+        command = command.split()
 
-                try:
-                    success = chosen_executor.execute(command)
-                except NoDiseaseInCityException as e:
-                    logging.error(e)
-                    success = False
-                if not success:
-                    logging.error(
-                        'Command cannot be executed. Type some other command.')
-                continue
-
-                logging.info(
-                    'Command successfully executed.')
+        for executor_class in COMMANDS:
+            executor = executor_class(self.game, self.current_player, self)
+            if executor.check_valid_command(command):
+                break
+        else:
+            logging.error(
+                'Command cannot be parsed. Type a correct command.')
+            return
 
+        try:
+            success = executor.execute(command)
+        except NoDiseaseInCityException as e:
+            logging.error(e)
+            success = False
+        if not success:
+            logging.error(
+                'Command cannot be executed. Type some other command.')
+
+        if self.current_player.action_count == 0:
             logging.info(
                 'No actions left. Now getting cards...')
-
-            # TODO: this must be a game method
-            for i in range(2):
-                game.draw_card(player)
-
+            self.end_turn()
+        else:
             logging.info(
-                'Cards drawn. Now starting infect phase.')
+                f'Actions left: {self.current_player.action_count}')
 
-            game.infect_city_phase()
+    def run(self):
```

Comment:
It's better to put this procedure at the top right after `__init__`. Generally, it's preferable to place most general functions at the top of class definition followed by functions that are called inside of the former.
My suggested order is:
`.run` --> `.start_game` --> `._switch player` --> `.run_single_command` --> `.end_turn`.

Reply:
OK. Got it.